evaluator.go

```diff
@@ -645,10 +692,47 @@ func (c *evalContext) addResult(value string) {
 	c.currentResult.Add(value)
 }
 
+func (c *evalContext) addSliceToResult(slice []interface{}) {
+	if c.currentResult.Cardinality()+len(slice) >= MaxResults {
+		panic(tooManyResults{})
+	}
+	for _, value := range slice {
+		c.currentResult.Add(value.(string))
+	}
+}
+
 func (c *evalContext) resultIter() <-chan interface{} {
 	return c.currentResult.Iter()
 }
 
 type evalNode interface {
 	visit(*State, *evalContext) error
 }
+
+func (state *State) getResultsFromCachedCQRforSlice(slice []interface{}) Result {
+	sliceMap := mapset.NewSetFromSlice(slice)
+	context := newContext()
+	// look in cachedCQR
+	for name := range sliceMap.Iter() {
+		if subResult, ok := state.cachedCQR.Get(name.(string)); ok {
+			//add subresults to context.currentResult
+			context.addSliceToResult(subResult.(Result).ToSlice())
+			// remove names we found in cache
+			sliceMap.Remove(name)
+		}
+	}
+	// return if we found all queries from cachedCQR
+	if sliceMap.Cardinality() == 0 {
+		return context.currentResult
+	}
+	// populate cachedCQR for leftOver elements in slice
+	state.populateCachedCQRforSlice(sliceMap.ToSlice())
```

Comment:
Pre populating cache would increase start up time, hence decided to populate cache as we query.  Need to add stats on how much time it takes to build cache and then will modify it accordingly